File: Q-1489 - Find Critical and Pseudo-Critical Edges in Minimum Spanning Tree/1489.py
class Solution:
    def findCriticalAndPseudoCriticalEdges(self, n: int, edges: List[List[int]]) -> List[List[int]]:
        adj = [[] for i in range(n)]
        parent = [i for i in range(n)]
        tree = list()        
        critical = list()
        notcritical = list()
        level = [1] * n 
        
        def find(i):
            if(parent[i] == i):
                return i
            parent[i] = find(parent[i])
            return parent[i]

        def union(pu, pv):
            if(level[pu] > level[pv]):
                parent[pv] = pu
            elif(level[pu] < level[pv]):
                parent[pu] = pv
            else:
                parent[pv] = pu
                level[pv] += 1

        def path(adj, nodeweight, v, stack, vis):
            u = nodeweight[0]
            vis[u] = 1
            stack.append(nodeweight)
            if u == v:
                return True
            for child in adj[u]:
                if not(vis[child[0]]):
                    if(path(adj, child, v, stack, vis)):
                        return True
            stack.pop(-1)

        edges = [(u, v, w, i) for i, (u,v,w) in enumerate(edges)]
        edges.sort(key = lambda x: x[2])
        e = len(edges)
        count = 0
        i = 0
        while(i < e):
            if(count == n-1 and edges[i][2] > last): 
                break
            (u, v, w, ind) = edges[i]
            urep = find(u)
            vrep = find(v)
            if(urep != vrep):
                adj[u].append((v, w, ind))
                adj[v].append((u, w, ind))
                tree.append(ind)
                last = w
                count += 1
                union(urep, vrep)
            else:
                if(last == w):
                    stack = list()
                    vis = [0] * n
                    path(adj, (u,0,n) , v, stack, vis) 
                    for half in stack:
                        if(half[1] == w):
                            if(half[2] not in notcritical):
                                notcritical.append(half[2]) 
                            if(ind not in notcritical):
                                notcritical.append(ind)        
            i+= 1
        for i in tree:
            if(i not in notcritical):
                critical.append(i)
        return [critical, notcritical]

File: Q-1489 - Find Critical and Pseudo-Critical Edges in Minimum Spanning Tree/1489.cpp
class Solution {
public:
    static bool compare(vector<int>&v1, vector<int>&v2){
        return v1[2] < v2[2];
    }
    vector<int> vec;
    int find(int x){
        if(vec[x] == -1){
            return x;
        }
        return find(vec[x]);
    }
	

    int tree(vector<vector<int>> &edges,vector<int> &in,vector<int> &ex){     
        int x = 0;
        if(in.size()){
            vec[in[0]] = in[1];
            x += in[2];
        }
        for(auto i:edges){
            if(i == ex){
                continue;
            }
            int item1 = find(i[0]);
            int item2 = find(i[1]);
            if(item1 != item2){
                vec[item1] = item2;
                x += i[2];
            }
        }
        
        int no_of_vec = 0;       
        for(auto i:vec){
            if(i == -1){
                no_of_vec++;
            }
        }
        if(no_of_vec>1)
        {
            return INT_MAX;
        }
        return x;
    }
    
    vector<vector<int>> findCriticalAndPseudoCriticalEdges(int n, vector<vector<int>>& edges) {
        for(int i = 0; i < n; i++) {
            vec.push_back(-1);   
        }
        for(auto i = 0; i < edges.size(); i++){
            edges[i].push_back(i);
        }
        sort(edges.begin(), edges.end(), compare);
        vector<int> temp;
        vector<vector<int>> answer(2);
        int min = tree(edges,temp,temp);

        for(auto i:edges) {                            
            for(int j = 0;j < n; j++) {
                vec[j] = -1;
            }
            int ec = tree(edges,temp,i);
            if(ec > min) {
                answer[0].push_back(i[3]);
            }
            else {
                for(int j=0;j<n;j++) {
                    vec[j]=-1;
                }
                int ic=tree(edges,i,temp);
                if(ic==min)
                {
                    answer[1].push_back(i[3]);
                }
            }
        }
        return answer;
    }
};